lib/sprinkler_web/live/device_live/index.ex

```diff
@@ -0,0 +1,52 @@
+defmodule SprinklerWeb.DeviceLive.Index do
+  use SprinklerWeb, :live_view
+
+  alias Sprinkler.Devices
+  alias Sprinkler.Devices.Device
+
+  @impl true
+  def mount(_params, _session, socket) do
+    {:ok, assign(socket, :devices, list_devices())}
+  end
+
+  @impl true
+  def handle_params(params, _url, socket) do
+    {:noreply, apply_action(socket, socket.assigns.live_action, params)}
+  end
+
+  defp apply_action(socket, :edit, %{"id" => id}) do
+    socket
+    |> assign(:page_title, "Edit Device")
+    |> assign(:device, Devices.get_device!(id))
+  end
+
+  defp random_string do
```

Comment:
We can send this function to the bottom of the file just no to break the implementation of `apply_action/3`, it's weird not to see all partial implementations together.